### PR TITLE
Infer 'isolated' closure parameters from context.

### PR DESCRIPTION
**Pull request by DougGregor**
When a closure is provided with a contextual type that has isolated
parameters, infer that the corresponding closure parameter is "isolated".

Fixes rdar://83732479.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2360,6 +2363,9 @@ class ConstraintSystem {
   llvm::SmallMapVector<const CaseLabelItem *, CaseLabelItemInfo, 4>
       caseLabelItems;
 
+  /// The set of parameters that have been inferred to be 'isolated'.
+  llvm::SmallVector<ParamDecl *, 2> isolatedParams;
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply include/swift/Sema/ConstraintSystem.h:2367]: Sure!

**Review by ktoso — APPROVED** [review]
Awesome, LGTM.

Curious why we prepare this for multiple isolated params, AFAIR we restrict to one nowadays -- just future proofing I guess?

**Comment by DougGregor** [comment]
@swift-ci please test Windows

File: lib/Sema/CSSolver.cpp [hunk 2]
```diff
@@ -186,6 +186,8 @@ Solution ConstraintSystem::finalize() {
 
   solution.solutionApplicationTargets = solutionApplicationTargets;
   solution.caseLabelItems = caseLabelItems;
+  solution.isolatedParams.clear();
```
Comment (xedin) [inline lib/Sema/CSSolver.cpp:189]: Nit: I don't think we need a to clear here because it's a brand new solution allocated by `finalize`.

**Review by xedin — APPROVED** [review]

File: lib/Sema/CSSolver.cpp (hunk 2 above)
Comment (DougGregor) [reply lib/Sema/CSSolver.cpp:189]: Sure, I'll drop it in a follow-up commit

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**